### PR TITLE
BB-425 Short version IDs break ordering - use lastModified

### PR DESCRIPTION
**Pull request by nicolas2bert**
**Summary:**

- Short version IDs no longer preserve ordering.
- We will use the "last-modified" date to sort version and delete markers.

**Description:**

- Prior to the introduction of short version IDs, a lower version ID indicated a more recent version. However, with short version IDs, the ordering is no longer preserved.
- To address this issue, we will use the "last-modified" date to sort version and delete markers from the object versions listing. This will ensure that the versions are always sorted in chronological order, regardless of the length of the version ID.
- We considered decoding the version ID as a solution, but this would be quite expensive in terms of CPU time and memory usage.

**Encoding algorithms history:**

The "short version id" encoding uses lossy base62 encoding algorithms (Base62Integer, Base62String) that do not preserve the order of the timestamps in the version ID. Previously, encoding was done using the hex encoding algorithm, which is a lossless encoding algorithm because it does not lose any information during the encoding process.

### CODE REVIEW

**Review by rahulreddy — COMMENTED** [review]
@nicolas2bert Why not decode the version id as they are guaranteed to be ordered?

**Comment by nicolas2bert** [comment]
The "base62Decode" function ([Arsenal/lib/versioning/VersionID.ts at development/7.10 · scality/Arsenal](https://github.com/scality/Arsenal/blob/development/7.10/lib/versioning/VersionID.ts#L215) ) could be quite expensive in terms of CPU time and memory usage. There are a few reasons for this:

- The function makes three calls to the base62Integer.decode() and base62String.decode() functions. These functions are themselves relatively expensive, so making three calls to them adds up.

- The function has to iterate through the string to find the start of each encoded value. This is also an expensive operation.

- The function has to slice the string three times. This is another relatively expensive operation.

All of these factors add up to make this function relatively expensive.

**Comment by nicolas2bert** [comment]
@rahulreddy  Is there any advantage to using version ID over creation date?

The only disadvantage, I see, of using the creation date is that two versions could be created at the same time.
In that case, we could decode the version ID to decide.

**Comment by rahulreddy** [comment]
> @rahulreddy Is there any advantage to using version ID over creation date?
> 
> The only disadvantage, I see, of using the creation date is that two versions could be created at the same time. In that case, we could decode the version ID to decide.

Yup, my main concern is the collision. You will have two versions created at the same ms, we already see at at a lot of deployments today. I would add another layer of entropy to be unique, may be you decode version ids and compare only for this case.

**Comment by nicolas2bert** [comment]
Yes, I agree. This condition is handled in the PR.

**Comment by francoisferrand** [comment]
AFAICT, we also rely on ordering of keys in MongoClientInterface (and possibly metadata ?) to compute the last "version" to use as master... Does it mean we will need to do the same kind of fix in Arsenal : https://github.com/scality/Arsenal/blob/bdb59a0e63c20578078bd169e70c78a0c49f6e60/lib/storage/metadata/mongoclient/MongoClientInterface.js#L1055 ?

**Comment by alexanderchan-scality** [comment]
do correct me if I am wrong, short version ID is only used at the encoding/decoding step in cloudserver, and the `generateVersionID` (https://github.com/scality/Arsenal/blob/bdb59a0e63c20578078bd169e70c78a0c49f6e60/lib/versioning/VersionID.ts#L100) is still timestamp based, so list order of keys should still be in some chronological order; the description may be a bit misleading attributing the issue to shortVersionID.

this issue would then look like an edge-case (race condition?)  scenario that will also affect the long version ID and would likely cause issue to the `getLatestVersion` logic pointed out by  Francois

**Comment by nicolas2bert** [comment]
Internal medatada methods (server side) use decoded version id. For decoded version ids, the order is preserved. 

However, the lifecycle "merging and sorting" (client side) method used base62 encoded version id to sort the listing (with merging versions and delete markers) that does not seem to preserve the order (based on creation date). Cf PR description. 


I am not sure to understand your concerns @francoisferrand @alexanderchan-scality .

**Comment by alexanderchan-scality** [comment]
@nicolas2bert i see. I had confused the versionID comparison to be with the internal version id; then I do agree with you that the base62 encode does lose the lexigraphic order property needed for the comparison. 

In that case, the find latest version would not be affected since that uses the internal VID for comparison.

**Comment by francoisferrand** [comment]
> Internal medatada methods (server side) use decoded version id. For decoded version ids, the order is preserved.

thanks @nicolas2bert , that is the part I missed, and got me worried of similar issue with arsenal.

s3 listObjectVersions is supposed to "_request returns objects in the order that they were stored, returning the most recently stored object first_". I do not see any mention (in the doc) that this order covers both versions & delete markers, but it seems to be consistent with the above statement and the exemples there...

So I wonder why we do this sorting in lifecycle, and if we should not implement this in cloudserver/arsenal instead? (This may avoid issues in the UI as well)

**Comment by nicolas2bert** [comment]
@francoisferrand The listObjectVersions API returns versions and delete markers in separate arrays. To determine the object's "stale date," the arrays must be merged and sorted.
With the introduction of lifecycle optimization listings, this logic will be deprecated.

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 1]
```diff
@@ -453,13 +458,42 @@ class LifecycleTask extends BackbeatTask {
...
     }
 
+    /** _decodeVID - decode the version id
+     * @param {string} versionId - version ID
+     * @param {Logger.newRequestLogger} log - logger object
+     * @return {Object} result - { error, versionId }
+     * @return {Error} result.error - if decoding failed
+     * @return {String} result.versionId - decoded version id
+     */
+    _decodeVID(versionId, log) {
```
Comment (francoisferrand) [inline extensions/lifecycle/tasks/LifecycleTask.js:468]: is this the same as arsenal's `versioning/decode` ?

**Review by alexanderchan-scality — APPROVED** [review]

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 2]
```diff
@@ -483,40 +517,57 @@ class LifecycleTask extends BackbeatTask {
...
-                        // 3. choose one randomly since all conditions match
-                        // TODO: to be fixed
-                        sortedList.push(versions[vIdx++]);
-                    }
+                // The `Key` names of the versions and delete markers are the same.
+                // Compare the `LastModified` timestamps of the versions and delete markers.
+                const versionLastModified = new Date(versions[vIdx].LastModified);
+                const deleteMarkerLastModified = new Date(deleteMarkers[dmIdx].LastModified);
+                const isVersionLastModifiedNewer = versionLastModified > deleteMarkerLastModified;
+                const isDMLastModifiedNewer = deleteMarkerLastModified > versionLastModified;
```
Comment (francoisferrand) [inline extensions/lifecycle/tasks/LifecycleTask.js:525]: nit: this makes the code harder to read IMO, as it "hides" the real intent here... i.e. just comparing the date, and handling the 3 possible outcomes of the comparison....

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 3]
```diff
@@ -483,40 +517,57 @@ class LifecycleTask extends BackbeatTask {
...
+                if (isVersionLastModifiedNewer) {
+                    sortedList.push(versions[vIdx++]);
+                } else if (isDMLastModifiedNewer) {
+                    sortedList.push(deleteMarkers[dmIdx++]);
                 } else {
-                    // 4. by VersionId, lower number means newer
-                    if (isVersionVidNewer) {
+                    // If the version and the delete marker have the same last modified date
+                    const nullVersion = (versions[vIdx].VersionId === 'null'
+                    || deleteMarkers[dmIdx].VersionId === 'null');
```
Comment (francoisferrand) [inline extensions/lifecycle/tasks/LifecycleTask.js:534]: nit: missing indent

File: extensions/lifecycle/tasks/LifecycleTask.js [hunk 4]
```diff
@@ -483,40 +517,57 @@ class LifecycleTask extends BackbeatTask {
...
+                        }
+
+                        const { error: decodedDMError, decodedVersionId: decodedDMId } =
+                            this._decodeVID(deleteMarkers[dmIdx].VersionId, log);
+                        if (decodedDMError) {
+                            return { error: decodedDMError, sortedList: null };
+                        }
+
+                        const isVersionVidNewer = decodedVersionId < decodedDMId;
+                        if (isVersionVidNewer) {
```
Comment (francoisferrand) [inline extensions/lifecycle/tasks/LifecycleTask.js:560]: these push to `sortedList` and incrementation are repeated in multiple places, while they are actually the critical step of the algorithm.

to make core safer & easier to test, would be better to introduce a `bool compare(version, deleteMarker)` function, which would only perform the comparison (and be easy to test all corner cases), and keep the code simple & readable in the algorithm (handling only the progression and update of the list, ie. all `sortedList.push(XXXXX[YYYYY++])`)

**Review by francoisferrand — APPROVED** [review]

**Comment by nicolas2bert** [comment]
@bert-e approve

**Comment by nicolas2bert** [comment]
ping
